### PR TITLE
examples: Update MPI tutorial notebook and scripts

### PR DESCRIPTION
**Pull request by georgebisbas**
Update the setup script and notebook.
Add deterministic output (--no-stream)

Much needed after 3 years.

### CODE REVIEW

File: scripts/create_ipyparallel_mpi_profile.sh [hunk 1]
```diff
@@ -6,11 +6,6 @@ IPYTHONDIR=~/.ipython
 # Create a new profile, called "mpi"
 ipython profile create --parallel --profile=mpi
 
-# Add the following line as per instructions from
```
Comment (FabioLuporini) [inline scripts/create_ipyparallel_mpi_profile.sh:9]: out of curiosity, why are you dropping this? unnecessary?

**Comment by FabioLuporini** [comment]
Thank you. So, you've tested it locally, right? and it still does work? nice

File: scripts/create_ipyparallel_mpi_profile.sh (hunk 1 above)
Comment (georgebisbas) [reply scripts/create_ipyparallel_mpi_profile.sh:9]: Yes, the machinery in line 12 has been deprecated.
The option engines=mpi I am using here:

https://github.com/devitocodes/devito/pull/1923/files#diff-4acc409654fcff44dde5c96e156d25a86f29d10a612d110f245af6f76a47a443R25

is taking care of instructing the MPI launchers.

**Review by georgebisbas — COMMENTED** [review]
Maybe this can be merged if happy

**Comment by georgebisbas** [comment]
> Thank you. So, you've tested it locally, right? and it still does work? nice

Tested for me and alongside Kene as well

**Comment by mloubout** [comment]
Can it be re-enable in the ci or is new version still notnworking

File: scripts/create_ipyparallel_mpi_profile.sh [hunk 2]
```diff
@@ -6,13 +6,9 @@ IPYTHONDIR=~/.ipython
...
-# https://ipyparallel.readthedocs.io/en/latest/process.html#using-ipcluster-in-mpiexec-mpirun-mode 
-# This is instructing `ipcluster` to use the MPI launchers
-echo "c.IPClusterEngines.engine_launcher_class = 'MPIEngineSetLauncher'" >> $IPYTHONDIR/profile_mpi/ipcluster_config.py
-
 ver=$(mpiexec --version)
 if [[ $ver == *"open-mpi"* ]]; then
   # OpenMPI need to be told that is allowed to oversubscribe cores
+  rm $IPYTHONDIR/profile_mpi/ipcluster_config.py
```
Comment (georgebisbas) [inline scripts/create_ipyparallel_mpi_profile.sh:12]: Delete any previous profiles in host.

File: .github/workflows/examples-mpi.yml [hunk 3]
```diff
@@ -47,10 +47,13 @@ jobs:
 
     - name: Test mpi notebooks
       run : |
-        # Currently not tested due to issue #859
-        # ipcluster start --profile=mpi -n 4 --daemon
-        # py.test --nbval examples/mpi
-        # ipcluster stop --profile=mpi
+        pip3 install "ipyparallel<8.4"
```
Comment (FabioLuporini) [inline .github/workflows/examples-mpi.yml:50]: @georgebisbas out of curiosity, why <8.4 ? and shouldn't this rather be added to the "Install dependencies" section above?

File: .github/workflows/examples-mpi.yml [hunk 4]
```diff
@@ -43,14 +43,16 @@ jobs:
     - name: Install dependencies
       run: |
         pip install --upgrade pip
-        pip install -e .
+        pip install -e .[extras]
```
Comment (mloubout) [inline .github/workflows/examples-mpi.yml:46]: `[extras,mpi]`?

File: requirements-optional.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
 matplotlib
-ipyparallel
+ipyparallel<8.4
```
Comment (mloubout) [inline requirements-optional.txt:2]: move to mpi requirements

**Review by mloubout — COMMENTED** [review]
Almost there, two small comments

File: .github/workflows/examples-mpi.yml (hunk 4 above)
Comment (georgebisbas) [reply .github/workflows/examples-mpi.yml]: done

File: requirements-mpi.txt [hunk 6]
```diff
@@ -1 +1,2 @@
-mpi4py
+mpi4py<4.0
```
Comment (georgebisbas) [inline requirements-mpi.txt:1]: NFR: this was added because soon mpi4py will go to version 4 after long time of minor upgrades

Comment (mloubout) [reply requirements-mpi.txt:1]: Does dependabot actually understand this is a dependency file as well?

**Pull request merged**